src/lib/dnssd/minimal_mdns/core/RecordWriter.cpp

```diff
@@ -0,0 +1,186 @@
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include "RecordWriter.h"
+
+namespace mdns {
+namespace Minimal {
+
+SerializedQNameIterator RecordWriter::PreviousName(size_t index) const
+{
+    if (index >= kMaxCachedReferences)
+    {
+        return SerializedQNameIterator();
+    }
+
+    uint16_t offset = mPreviousQNames[index];
+    if (offset == kInvalidOffset)
+    {
+        return SerializedQNameIterator();
+    }
+
+    return SerializedQNameIterator(BytesRange(mOutput->Buffer(), mOutput->Buffer() + mOutput->WritePos()),
+                                   mOutput->Buffer() + offset);
+}
+
+chip::Optional<uint16_t> RecordWriter::FindPreviousName(const FullQName & name) const
+{
+    for (size_t i = 0; i < kMaxCachedReferences; i++)
+    {
+        SerializedQNameIterator previous = PreviousName(i);
+
+        // Any of the sub-segments may match
+        while (previous.IsValid())
+        {
+            if (previous == name)
+            {
+                return chip::Optional<uint16_t>::Value(previous.OffsetInCurrentValidData());
+            }
+
+            if (!previous.Next())
+            {
+                break;
+            }
+        }
+    }
+
+    return chip::Optional<uint16_t>::Missing();
+}
+
+chip::Optional<uint16_t> RecordWriter::FindPreviousName(const SerializedQNameIterator & name) const
+{
+    for (size_t i = 0; i < kMaxCachedReferences; i++)
+    {
+        SerializedQNameIterator previous = PreviousName(i);
+
+        // Any of the sub-segments may match
+        while (previous.IsValid())
+        {
+            if (previous == name)
+            {
+                return chip::Optional<uint16_t>::Value(previous.OffsetInCurrentValidData());
+            }
+
+            if (!previous.Next())
+            {
+                break;
+            }
+        }
+    }
+
+    return chip::Optional<uint16_t>::Missing();
+}
+
+void RecordWriter::WriteQName(const FullQName & qname)
+{
+    size_t qNameWriteStart = mOutput->WritePos();
+
+    for (uint16_t i = 0; i < qname.nameCount; i++)
+    {
+
+        // find out if the record part remaining already is located somewhere
+        FullQName remaining;
+        remaining.names     = qname.names + i;
+        remaining.nameCount = qname.nameCount - i;
+
+        // Try to find a valid offset
+        chip::Optional<uint16_t> offset = FindPreviousName(remaining);
+
+        if (offset.HasValue())
+        {
+            // Pointer to offset: set the highest 2 bits
+            mOutput->Put16(offset.Value() | 0xC000);
+
+            if (mOutput->Fit())
+            {
+                RememberWrittenQnameOffset(qNameWriteStart);
+            }
```

Comment:
Curious why we remember compressed QNAME elements rather than just the literal ones written in full

Reply:
I changed it to remember  things that are not fully compressed.
This remembers it because I may have "bar" then "foo.bar" (bar referenced) then when I compress "abc.foo.bar" I want to link to the longer "foo.bar" instead of just bar.